13_Parallel_programming_PR_for_review/download_single_process.py

```diff
@@ -0,0 +1,213 @@
+    """
+    Base class for arguments parsing
+    """
+
+    def __init__(
+        self, *args, formatter_class=argparse.RawTextHelpFormatter, **kwargs
+    ):
+        super(BaseArgumentParser, self).__init__(
+            *args, formatter_class=formatter_class, **kwargs
+        )
+
+
+def create_parser():
+    """
+    The function creates the arguments parser
+    :return: The arguments parser
+    """
+    parser = BaseArgumentParser(description=__doc__, epilog=EPILOG)
+    parser.add_argument(
+        "filename", nargs="+", help="set URL list file",
+    )
+    parser.add_argument(
+        "--dir",
+        dest="directory",
+        default=DEFAULT_DIRECTORY,
+        type=str,
+        help="set output directory",
+    )
+    parser.add_argument(
+        "--threads",
+        dest="threads",
+        default=DEFAULT_THREADS,
+        type=int,
+        help="set number of threads",
+    )
+    parser.add_argument(
+        "--size",
+        dest="size",
+        default=DEFAULT_SIZE,
+        type=str,
+        help="set output image sizes",
+    )
+
+    return parser
+
+
+def log_info(msg):
+    """
+    Log an info message to stdout
+    :param msg: message string
+    """
+    sys.stdout.write(f"[Info] {datetime.now()}: {msg}\n")
+    sys.stdout.flush()
+
+
+def log_error(msg):
+    """
+    Log an error message to stdout
+    :param msg: message string
+    """
+    sys.stdout.write(f"[Error] {datetime.now()}: {msg}\n")
+    sys.stdout.flush()
+
+
+def download_file(url):
+    """
+    Download file from URL
+    :param url: URL to source file
+    :return: file content if download succeed otherwise 1
+    """
+    try:
+        with urlopen(url) as response:
+            if response.status != 200:
+                log_error(
+                    f"Downloading from {url} failed: "
+                    f"{response.status} {response.reason}"
+                )
+                return 1
+
+            return response.read()
+
+    except Exception as exception:
+        log_error(exception)
+        return 1
+
+
+def scale_image(input_image, width=None, height=None):
+    """
+    The scaling method takes into account the aspect ratio of the picture.
+    :param input_image: input image
+    :param width: output image width
+    :param height: output image height
+    :return: modified image
+    """
+    w, h = input_image.size
+
+    if width and height:
+        max_size = (width, height)
+    elif width:
+        max_size = (width, h)
+    elif height:
+        max_size = (w, height)
+    else:
+        raise RuntimeError("Width or height required!")
+
+    input_image.thumbnail(max_size, Image.ANTIALIAS)
+
+    return input_image
+
+
+def output_directory(path):
+    """
+    # Create output directory
+    :param path: name output directory
+    :return: full path for output directory
+    """
+    output_dir = os.path.join(os.getcwd(), path)
+    if not os.path.exists(output_dir):
+        os.mkdir(output_dir, mode=0o777)
```

Comment:
Why to set mode?

Reply:
Done. Removed mode=0o777. I agree it is unnecessary here.